From 3cd40ff60509e9552edb3656596882714094f493 Mon Sep 17 00:00:00 2001
From: Andrew Woloszyn <andrew.woloszyn@gmail.com>
Date: Thu, 30 Apr 2026 12:10:15 -0400
Subject: [PATCH] Apply IREE patch 0001-hrx-pruned-runtime-deps

Reapplies the HRX-local patch stack on top of the new pristine import
(now pinned to users/awoloszyn/hrx_on_amdgpu, IREE a56a07822441). The
host_queue_command_buffer_test guard had to be hand-applied because the
test's SRCS list moved from host_queue_command_buffer_internal.h to
host_queue_command_buffer_packet.h on the new base.

Made-with: Cursor
---
 CMakeLists.txt                                | 39 ++++++++++++++-
 build_tools/cmake/iree_check_test.cmake       |  4 ++
 build_tools/cmake/iree_install_support.cmake  |  1 +
 .../iree/hal/drivers/amdgpu/CMakeLists.txt    | 48 ++++++++++---------
 .../src/iree/hal/drivers/amdgpu/abi/common.h  |  2 +-
 .../hal/drivers/amdgpu/util/CMakeLists.txt    |  4 ++
 .../hal/drivers/amdgpu/util/libaqlprofile.c   |  2 +-
 .../src/iree/hal/drivers/amdgpu/util/libhsa.h | 10 ++--
 8 files changed, 78 insertions(+), 32 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 24137ca..4b6d72f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -981,7 +981,25 @@ endif()
 
 if(IREE_BUILD_TESTS)
   iree_set_googletest_cmake_options()
-  add_subdirectory(third_party/googletest EXCLUDE_FROM_ALL)
+  find_package(GTest REQUIRED)
+  if(NOT TARGET gtest AND TARGET GTest::gtest)
+    add_library(gtest ALIAS GTest::gtest)
+  endif()
+  if(NOT TARGET gtest_main AND TARGET GTest::gtest_main)
+    add_library(gtest_main ALIAS GTest::gtest_main)
+  endif()
+  if(NOT TARGET gmock AND TARGET GTest::gmock)
+    add_library(gmock ALIAS GTest::gmock)
+  endif()
+  if(NOT TARGET gmock_main AND TARGET GTest::gmock_main)
+    add_library(gmock_main ALIAS GTest::gmock_main)
+  endif()
+  if(NOT TARGET gtest OR NOT TARGET gtest_main OR
+     NOT TARGET gmock OR NOT TARGET gmock_main)
+    message(FATAL_ERROR
+      "IREE_BUILD_TESTS requires external GTest targets: "
+      "GTest::gtest, GTest::gtest_main, GTest::gmock, and GTest::gmock_main")
+  endif()
 endif()
 
 if(IREE_ENABLE_THREADING)
@@ -1057,7 +1075,24 @@ if(IREE_ALLOCATOR_SYSTEM STREQUAL "mimalloc")
 endif()
 
 if(IREE_HAL_DRIVER_AMDGPU)
-  add_subdirectory(build_tools/third_party/hsa-runtime-headers EXCLUDE_FROM_ALL)
+  find_package(hsa-runtime64 CONFIG REQUIRED)
+  if(NOT TARGET hsa_runtime::headers)
+    get_target_property(_IREE_HSA_RUNTIME_INCLUDE_DIRS
+      hsa-runtime64::hsa-runtime64 INTERFACE_INCLUDE_DIRECTORIES)
+    if(NOT _IREE_HSA_RUNTIME_INCLUDE_DIRS)
+      message(FATAL_ERROR
+        "hsa-runtime64::hsa-runtime64 does not publish include directories")
+    endif()
+    add_library(hsa_runtime_headers INTERFACE)
+    target_include_directories(hsa_runtime_headers SYSTEM INTERFACE
+      ${_IREE_HSA_RUNTIME_INCLUDE_DIRS})
+    add_library(hsa_runtime::headers ALIAS hsa_runtime_headers)
+    iree_install_targets(
+      TARGETS hsa_runtime_headers
+      COMPONENT IREEBundledLibraries
+      EXPORT_SET Runtime
+    )
+  endif()
 endif()
 
 if(IREE_HAL_DRIVER_CUDA)
diff --git a/build_tools/cmake/iree_check_test.cmake b/build_tools/cmake/iree_check_test.cmake
index 96271e2..b588a8c 100644
--- a/build_tools/cmake/iree_check_test.cmake
+++ b/build_tools/cmake/iree_check_test.cmake
@@ -55,6 +55,10 @@ function(iree_check_test)
     ${ARGN}
   )
 
+  if(NOT TARGET iree-check-module)
+    return()
+  endif()
+
   # Normalize some variables before using them.
   string(TOUPPER ${_RULE_TARGET_BACKEND} _UPPERCASE_TARGET_BACKEND)
   string(REPLACE "-" "_" _NORMALIZED_TARGET_BACKEND ${_UPPERCASE_TARGET_BACKEND})
diff --git a/build_tools/cmake/iree_install_support.cmake b/build_tools/cmake/iree_install_support.cmake
index 2037452..7cdb6aa 100644
--- a/build_tools/cmake/iree_install_support.cmake
+++ b/build_tools/cmake/iree_install_support.cmake
@@ -126,6 +126,7 @@ function(iree_generate_export_targets)
   get_property(_export_targets GLOBAL PROPERTY "IREE_EXPORT_TARGETS_${_RULE_EXPORT_SET}")
   if(NOT _export_targets)
     message(STATUS "Skipping generation of export set ${_RULE_EXPORT_SET} (no targets matched)")
+    return()
   endif()
 
   set(_component "${_RULE_COMPONENT}")
diff --git a/runtime/src/iree/hal/drivers/amdgpu/CMakeLists.txt b/runtime/src/iree/hal/drivers/amdgpu/CMakeLists.txt
index 18f6ba9..02808f9 100644
--- a/runtime/src/iree/hal/drivers/amdgpu/CMakeLists.txt
+++ b/runtime/src/iree/hal/drivers/amdgpu/CMakeLists.txt
@@ -444,29 +444,31 @@ iree_cc_test(
     "iree-hal-drivers-amdgpu-tests"
 )
 
-iree_cc_test(
-  NAME
-    host_queue_command_buffer_test
-  SRCS
-    "host_queue_command_buffer_packet.h"
-    "host_queue_command_buffer_test.cc"
-  DEPS
-    ::amdgpu
-    ::headers
-    ::queue_affinity
-    iree::base
-    iree::hal
-    iree::hal::cts::util::test_base
-    iree::hal::drivers::amdgpu::cts::testdata_amdgpu
-    iree::hal::drivers::amdgpu::util::packet
-    iree::testing::gtest
-    iree::testing::gtest_main
-  LABELS
-    "driver=amdgpu"
-    "nodocker"
-  GROUP
-    "iree-hal-drivers-amdgpu-tests"
-)
+if(TARGET iree::hal::drivers::amdgpu::cts::testdata_amdgpu)
+  iree_cc_test(
+    NAME
+      host_queue_command_buffer_test
+    SRCS
+      "host_queue_command_buffer_packet.h"
+      "host_queue_command_buffer_test.cc"
+    DEPS
+      ::amdgpu
+      ::headers
+      ::queue_affinity
+      iree::base
+      iree::hal
+      iree::hal::cts::util::test_base
+      iree::hal::drivers::amdgpu::cts::testdata_amdgpu
+      iree::hal::drivers::amdgpu::util::packet
+      iree::testing::gtest
+      iree::testing::gtest_main
+    LABELS
+      "driver=amdgpu"
+      "nodocker"
+    GROUP
+      "iree-hal-drivers-amdgpu-tests"
+  )
+endif()
 
 iree_cc_test(
   NAME
diff --git a/runtime/src/iree/hal/drivers/amdgpu/abi/common.h b/runtime/src/iree/hal/drivers/amdgpu/abi/common.h
index bca7f1f..ffbecf1 100644
--- a/runtime/src/iree/hal/drivers/amdgpu/abi/common.h
+++ b/runtime/src/iree/hal/drivers/amdgpu/abi/common.h
@@ -61,7 +61,7 @@ typedef uint64_t uintptr_t;
 // HSA system type definitions. On the host side several abi/ types are
 // typedef'd directly to their HSA equivalents (e.g. iree_hsa_signal_t) so
 // that they can be used interchangeably with HSA API calls.
-#include "third_party/hsa-runtime-headers/include/hsa/hsa.h"  // IWYU pragma: export
+#include "hsa/hsa.h"  // IWYU pragma: export
 
 #endif  // IREE_AMDGPU_TARGET_DEVICE
 
diff --git a/runtime/src/iree/hal/drivers/amdgpu/util/CMakeLists.txt b/runtime/src/iree/hal/drivers/amdgpu/util/CMakeLists.txt
index 63dd650..aed9744 100644
--- a/runtime/src/iree/hal/drivers/amdgpu/util/CMakeLists.txt
+++ b/runtime/src/iree/hal/drivers/amdgpu/util/CMakeLists.txt
@@ -28,6 +28,10 @@ iree_cc_library(
   PUBLIC
 )
 
+# ROCm's packaged aqlprofile SDK header currently has C++-only declarations:
+# https://github.com/ROCm/rocm-systems/issues/5576
+set_source_files_properties("libaqlprofile.c" PROPERTIES LANGUAGE CXX)
+
 iree_cc_library(
   NAME
     libaqlprofile
diff --git a/runtime/src/iree/hal/drivers/amdgpu/util/libaqlprofile.c b/runtime/src/iree/hal/drivers/amdgpu/util/libaqlprofile.c
index 1ee198f..47e8c07 100644
--- a/runtime/src/iree/hal/drivers/amdgpu/util/libaqlprofile.c
+++ b/runtime/src/iree/hal/drivers/amdgpu/util/libaqlprofile.c
@@ -8,7 +8,7 @@
 
 #include "iree/base/internal/dynamic_library.h"
 #include "iree/base/internal/path.h"
-#include "third_party/hsa-runtime-headers/include/aqlprofile-sdk/aql_profile_v2.h"
+#include "aqlprofile-sdk/aql_profile_v2.h"
 
 // Keep vendor aqlprofile headers private to this translation unit. The public
 // wrapper mirrors only the narrow ABI surface used by the HAL so callers do not
diff --git a/runtime/src/iree/hal/drivers/amdgpu/util/libhsa.h b/runtime/src/iree/hal/drivers/amdgpu/util/libhsa.h
index c79e377..a44146e 100644
--- a/runtime/src/iree/hal/drivers/amdgpu/util/libhsa.h
+++ b/runtime/src/iree/hal/drivers/amdgpu/util/libhsa.h
@@ -15,11 +15,11 @@
 // TODO(benvanik): fork or upstream changes to the headers to make this safer?
 // Similar to Vulkan we could have all the function declarations hidden and also
 // include proper typed function pointers to aid in shared library use.
-#include "third_party/hsa-runtime-headers/include/hsa/amd_hsa_queue.h"  // IWYU pragma: export
-#include "third_party/hsa-runtime-headers/include/hsa/amd_hsa_signal.h"  // IWYU pragma: export
-#include "third_party/hsa-runtime-headers/include/hsa/hsa.h"  // IWYU pragma: export
-#include "third_party/hsa-runtime-headers/include/hsa/hsa_ext_amd.h"  // IWYU pragma: export
-#include "third_party/hsa-runtime-headers/include/hsa/hsa_ven_amd_loader.h"  // IWYU pragma: export
+#include "hsa/amd_hsa_queue.h"  // IWYU pragma: export
+#include "hsa/amd_hsa_signal.h"  // IWYU pragma: export
+#include "hsa/hsa.h"  // IWYU pragma: export
+#include "hsa/hsa_ext_amd.h"  // IWYU pragma: export
+#include "hsa/hsa_ven_amd_loader.h"  // IWYU pragma: export
 
 #ifdef __cplusplus
 extern "C" {
-- 
2.34.1

